parser-typechecker/src/Unison/Codebase/Execute.hs

```diff
@@ -0,0 +1,69 @@
+{-# LANGUAGE PartialTypeSignatures #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | Execute a computation of type '{IO} () that has been previously added to
+-- the codebase, without setting up an interactive environment.
+--
+-- This allows one to run standalone applications implemented in the Unison
+-- language.
+
+module Unison.Codebase.Execute where
+
+import Unison.Prelude
+
+import qualified Data.Map                      as Map
+
+import qualified Unison.Codebase               as Codebase
+import           Unison.Parser                  ( Ann )
+import qualified Unison.Parser                 as Parser
+import qualified Unison.Codebase.Runtime       as Runtime
+import           Unison.Codebase.Runtime       (Runtime)
+import qualified Unison.Term                   as Term
+import qualified Unison.UnisonFile             as UF
+import           Unison.Var                     ( Var )
+import qualified Unison.Var                    as Var
+import qualified Unison.PrettyPrintEnv         as PPE
+import qualified Unison.DataDeclaration        as DD
+import qualified Unison.HashQualified          as HQ
+import qualified Unison.Referent               as Referent
+import qualified Unison.Names3                 as Names3
+import qualified Unison.Codebase.Branch        as Branch
+import qualified Unison.Typechecker as Typechecker
+import           Unison.Runtime.IOSource       (nullaryMain)
+import           System.Exit (die)
+import           Control.Exception (finally)
+
+execute
+  :: Var v
+  => Codebase.Codebase IO v Ann
+  -> Runtime v
+  -> String
+  -> IO ()
+execute codebase runtime mainName =
+  (`finally` Runtime.terminate runtime) $ do
+    root <- Codebase.getRootBranch codebase
+    let parseNames0 = Names3.makeAbsolute0 (Branch.toNames0 (Branch.head root))
+    case HQ.fromString mainName of
+      Nothing -> die ("Not a function name: " ++ mainName)
+      Just hq -> do
+        let refs = Names3.lookupHQTerm hq (Names3.Names parseNames0 mempty)
+        case toList refs of
+          [Referent.Ref ref] -> do
+            typ <- Codebase.getTypeOfTerm codebase ref
+            let a = Parser.External
+                ppe = PPE.PrettyPrintEnv (const Nothing) (const Nothing)
+            case typ of
+              Just typ | Typechecker.isSubtype typ (nullaryMain a) -> do
+                let tm = DD.forceTerm a a (Term.ref a ref)
+                    codeLookup = Codebase.toCodeLookup codebase
+                    uf = UF.UnisonFile mempty mempty mempty
+                           (Map.singleton UF.RegularWatch [(Var.nameds "result", tm)])
+                selfContained <- Codebase.makeSelfContained' codeLookup uf
+                _ <- Runtime.evaluateWatches codeLookup ppe Runtime.noCache runtime selfContained
+                return ()
+              _ -> die (mainName ++ " is not of type '{IO} ()")
+          _ -> die ("Not found: " ++ mainName)
```

Comment:
Hmm so I guess this is a bit of a copy paste (but not exactly) of the existing code. How do we feel about this?

Other option might be to just add a flag to the existing codepath to control whether to print out the welcome, etc.

Reply:
The existing codepath does much more than printing the welcome message. It creates unneeded mutable references:

```
    rootRef                  <- newIORef root
    pathRef                  <- newIORef initialPath
    initialInputsRef         <- newIORef initialInputs
    numberedArgsRef          <- newIORef []
```

It starts watching for configuration, file system, and branch updates:

```
    (config, cancelConfig)   <-
      catchIOError (watchConfig configFile) $ \_ ->
        die "Your .unisonConfig could not be loaded. Check that it's correct!"
    cancelFileSystemWatch    <- watchFileSystem eventQueue dir
    cancelWatchBranchUpdates <- watchBranchUpdates (Branch.headHash <$>
                                                      readIORef rootRef)
                                                   eventQueue
                                                   codebase
```

It uses a queue to drive an event loop:

```
  eventQueue <- Q.newIO
      awaitInput = do
        -- use up buffered input before consulting external events
        i <- readIORef initialInputsRef
        case i of
          h:t -> writeIORef initialInputsRef t >> pure h
          [] ->
            -- Race the user input and file watch.
            Async.race (atomically $ Q.peek eventQueue) getInput >>= \case
              Left _ -> Left <$> atomically (Q.dequeue eventQueue)
              x      -> pure x
```

And then there are some free monad to organize all of that.

How would one hide all of that behind a flag? A little bit of duplication is better, in my opinion. If we want to remove duplication, the solution would be not a flag, but a proper abstraction that would be used both by this runner and the interactive loop.

Reply:
Also - appreciate this comprehensive reply that saved me from needing to dig into the code (that I don't remember super well). :)